clients/venice-admin-tool/src/main/java/com/linkedin/venice/TopicMessageFinder.java

```diff
@@ -143,4 +150,93 @@ public static byte[] serializeKey(String keyString, String keySchemaStr) {
 
     return keySerializer.serialize(key);
   }
+
+  protected static KeyPartitionInfo findPartitionIdForKey(
+      ControllerClient controllerClient,
+      String storeName,
+      int versionNumber,
+      String key,
+      String keySchemaStr) {
+    StoreResponse storeResponse = controllerClient.getStore(storeName);
+    if (storeResponse == null) {
+      throw new VeniceNoStoreException("Store " + storeName + " does not exist.");
+    }
+    StoreInfo storeInfo = storeResponse.getStore();
+    if (storeInfo == null) {
+      throw new VeniceNoStoreException("Store " + storeName + " does not exist.");
+    }
+
+    Optional<Version> versionInfo = storeInfo.getVersion(versionNumber);
+
+    int partitionCount;
+    PartitionerConfig partitionerConfig;
+    if (versionInfo.isPresent()) {
+      Version version = versionInfo.get();
+      LOGGER.info(
+          "Found store: {} version: {}. Will use partitioner config: {} and partition count: {} from this version",
+          storeName,
+          versionNumber,
+          version.getPartitionerConfig(),
+          version.getPartitionCount());
+      partitionerConfig = version.getPartitionerConfig();
+      partitionCount = version.getPartitionCount();
+    } else {
+      LOGGER.info(
+          "Store: {} version: {} not found. Will use partitioner config: {} from store level config",
+          storeName,
+          versionNumber,
+          storeInfo.getPartitionerConfig());
+      partitionerConfig = storeInfo.getPartitionerConfig();
+      partitionCount = storeInfo.getPartitionCount();
+    }
+
+    if (partitionCount <= 0) {
+      LOGGER.error("Partition count for store: {} is not set.", storeName);
+      throw new VeniceException("Partition count for store: " + storeName + " is not set.");
+    }
+
+    Properties params = new Properties();
+    params.putAll(partitionerConfig.getPartitionerParams());
+    VeniceProperties partitionerProperties = new VeniceProperties(params);
+    VenicePartitioner partitioner =
+        PartitionUtils.getVenicePartitioner(partitionerConfig.getPartitionerClass(), partitionerProperties);
+    LOGGER.info("The key schema for store: {} : {}", storeName, keySchemaStr);
+
+    byte[] keyBytes = TopicMessageFinder.serializeKey(key, keySchemaStr);
+    int partitionId = partitioner.getPartitionId(keyBytes, partitionCount);
+    System.out.println("Partition ID for key: " + key + " in store: " + storeName + " is: " + partitionId);
+    LOGGER.info("Partition ID for key: {} in store: {} is: {}", key, storeName, partitionId);
+
+    return new KeyPartitionInfo(storeInfo, keyBytes, partitionId, partitionCount);
+  }
+
+  static class KeyPartitionInfo {
```

Comment:
Should this be `protected` if it is being exposed as a return type of a `protected` function